crypto/storage/azure/keyvault.go

```diff
@@ -0,0 +1,276 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package azure
+
+import (
+	"context"
+	"crypto"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/security/keyvault/azkeys"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/crypto/storage/spi"
+	"golang.org/x/crypto/cryptobyte"
+	"golang.org/x/crypto/cryptobyte/asn1"
+	"io"
+	"net/http"
+	"regexp"
+	"time"
+)
+
+// New creates a new Azure Key Vault storage backend.
+// If useHSM is true, the key type will be azkeys.KeyTypeECHSM, otherwise azkeys.KeyTypeEC.
+func New(keyVaultUrl string, timeout time.Duration, useHSM bool) (spi.Storage, error) {
+	if keyVaultUrl == "" {
+		return nil, errors.New("missing Azure Key Vault URL")
+	}
+	cred, _ := azidentity.NewDefaultAzureCredential(nil)
+	client, err := azkeys.NewClient(keyVaultUrl, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Azure Key Vault client: %w", err)
+	}
+	return &keyvault{client: client, timeOut: timeout, useHSM: useHSM}, nil
+}
+
+// StorageType is the name of this storage type, used in health check reports and configuration.
+const StorageType = "azure-keyvault"
+
+type keyvault struct {
+	client  keyVaultClient
+	timeOut time.Duration
+	useHSM  bool
+}
+
+func (a keyvault) Name() string {
+	return StorageType
+}
+
+func (a keyvault) CheckHealth() map[string]core.Health {
+	return nil
+}
+
+func (a keyvault) NewPrivateKey(ctx context.Context, namingFunc func(crypto.PublicKey) (string, error)) (crypto.PublicKey, string, error) {
+	keyID, err := namingFunc(nil)
+	if err != nil {
+		return nil, "", err
+	}
+	// Make sure it doesn't already exist: Azure Key Vault otherwise creates a new version for the same key.
+	exists, err := a.PrivateKeyExists(ctx, keyID)
+	if err != nil {
+		return nil, "", err
+	}
+	if exists {
+		return nil, "", spi.ErrKeyAlreadyExists
+	}
+
+	keyName := keyIDToKeyName(keyID)
+
+	var keyType azkeys.KeyType
+	if a.useHSM {
+		keyType = azkeys.KeyTypeECHSM
+	} else {
+		keyType = azkeys.KeyTypeEC
+	}
+
+	response, err := a.client.CreateKey(ctx, keyName, azkeys.CreateKeyParameters{
+		Kty:   to.Ptr(keyType),
+		Curve: to.Ptr(azkeys.CurveNameP256),
+		KeyAttributes: &azkeys.KeyAttributes{
+			Enabled:    to.Ptr(true),
+			Exportable: to.Ptr(false),
+		},
+	}, nil)
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to create key in Azure Key Vault (name=%s): %w", keyName, err)
+	}
+	publicKey, _, err := parseKey(response.Key)
+	if err != nil {
+		return nil, "", err
+	}
+	return publicKey, keyID, nil
+}
+
+func (a keyvault) GetPrivateKey(ctx context.Context, kid string) (crypto.Signer, error) {
```

Comment:
this no longer provides direct access to the key. change to `Signer` to follow go convention?

Reply:
?

Reply:
We could do this, but out of scope.